node/libs/storage/src/block_store.rs

```diff
@@ -0,0 +1,214 @@
+//! Defines storage layer for finalized blocks.
+use anyhow::Context as _;
+use std::{collections::VecDeque, fmt, sync::Arc};
+use zksync_concurrency::{ctx, sync};
+use zksync_consensus_roles::validator;
+
+/// State of the `BlockStore`: continuous range of blocks.
+#[derive(Debug, Clone)]
+pub struct BlockStoreState {
+    /// Stored block with the lowest number.
+    pub first: validator::CommitQC,
+    /// Stored block with the highest number.
+    pub last: validator::CommitQC,
+}
+
+impl BlockStoreState {
+    /// Checks whether block with the given number is stored in the `BlockStore`.
+    pub fn contains(&self, number: validator::BlockNumber) -> bool {
+        self.first.header().number <= number && number <= self.last.header().number
+    }
+
+    /// Number of the next block that can be stored in the `BlockStore`.
+    /// (i.e. `last` + 1).
+    pub fn next(&self) -> validator::BlockNumber {
+        self.last.header().number.next()
+    }
+}
+
+/// Storage of a continuous range of L2 blocks.
+///
+/// Implementations **must** propagate context cancellation using [`StorageError::Canceled`].
+#[async_trait::async_trait]
+pub trait PersistentBlockStore: fmt::Debug + Send + Sync {
+    /// Range of blocks avaliable in storage. None iff storage is empty.
+    /// Consensus code calls this method only once and then tracks the
+    /// range of avaliable blocks internally.
+    async fn state(&self, ctx: &ctx::Ctx) -> ctx::Result<Option<BlockStoreState>>;
+
+    /// Gets a block by its number. Returns None if block is missing.
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: validator::BlockNumber,
+    ) -> ctx::Result<Option<validator::FinalBlock>>;
+
+    /// Persistently store a block.
+    /// Implementations are only required to accept a block directly after the current last block,
+    /// so that the stored blocks always constitute a continuous range.
+    /// Implementation should return only after the block is stored PERSISTENTLY -
+    /// consensus liveness property depends on this behavior.
+    async fn store_next_block(
+        &self,
+        ctx: &ctx::Ctx,
+        block: &validator::FinalBlock,
+    ) -> ctx::Result<()>;
+}
+
+#[derive(Debug)]
+struct Inner {
+    queued: sync::watch::Sender<BlockStoreState>,
+    persisted: BlockStoreState,
+    queue: VecDeque<validator::FinalBlock>,
```

Comment:
As per https://github.com/matter-labs/era-consensus/pull/53#discussion_r1441540001 I've moved the responsibility of keeping cache bounded to the user, so that both cache size limit and inflight request limit are enforced in the same place.

Reply:
added metrics